R/pool.R

```diff
@@ -45,118 +53,178 @@ Pool <- R6Class("Pool",
       }
       reg.finalizer(self,
         function(self) {
-          protectDefaultScheduler({
-            freeEnv <- private$freeObjects
-            if (length(freeEnv) > 0) {
-              self$close()
-            }
-          })
+          if (self$valid) {
+            warning("Closing leaked pool.")
+            self$close()
+          }
         },
         onexit = TRUE)
     },
+
     ## calls activate and returns an object
-    fetch = function() {                       #### WHAT TO DO IF THIS THROWS AN ERROR???
-      protectDefaultScheduler({
+    fetch = function() {
+      if (!self$valid) {
+        stop("This pool is no longer valid. Cannot fetch new objects.")
+      }
+      naiveScheduler$protect({
         if (self$counters$free + self$counters$taken >= self$maxSize) {
           stop("Maximum number of objects in pool has been reached")
         }
 
-        freeEnv <- private$freeObjects
         ## see if there's any free objects
+        freeEnv <- private$freeObjects
         if (length(freeEnv) > 0) {
-          ## get first free object we find
-          id <- ls(freeEnv)[[1]]
+          id <- ls(freeEnv)[[1]]  ## get first free object we find
           object <- freeEnv[[id]]
-
-          taskHandle <- attr(object, "reapTaskHandle", exact = TRUE)
-          if (!is.null(taskHandle)) {
-            attr(object, "reapTaskHandle") <- NULL
-            # Cancel the reap task.
-            taskHandle()
-          }
+          private$cancelReapTask(object)  ## cancel reap task if it exists
 
         } else {
           ## if we get here, there are no free objects
           ## and we must create a new one
           object <- private$createObject()
-          id <- attr(object, "id", exact = TRUE)
-        }
-        ## activate object and return it
-        private$changeObjectStatus(id, object, "free", "taken")
-        onActivate(object)
-        if (!onValidate(object)) {
-          stop("Object activation was not successful")
+          id <- attr(object, "..metadata", exact = TRUE)$..id
         }
 
+        ..metadata <- attr(object, "..metadata", exact = TRUE)
+        naiveScheduler$protect({
```

Comment:
Right... I think we can get rid of the outer `protect`, right? (I think we're only scheduling things inside the inner `protect`)